src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/UIHintAttribute.cs

```diff
@@ -187,21 +181,44 @@ public override bool Equals([NotNullWhen(true)] object? obj)
                     {
                         throw new InvalidOperationException(SR.Format(SR.UIHintImplementation_ControlParameterKeyIsNotAString,
                                                             i,
-                                                            inputControlParameters[i]!.ToString()));
+                                                            key.ToString()));
                     }
 
-                    if (controlParameters.ContainsKey(keyString))
+                    if (!controlParameters.TryAdd(keyString, value))
                     {
                         throw new InvalidOperationException(SR.Format(SR.UIHintImplementation_ControlParameterKeyOccursMoreThanOnce,
                                                             i,
                                                             keyString));
                     }
-
-                    controlParameters[keyString] = value;
                 }
 
                 return controlParameters;
             }
+
+            private static bool ControlParametersAreEqual(Dictionary<string, object?> left, Dictionary<string, object?> right)
+            {
```

Comment:
We should assert that the comparers are what we expect them to be.  This implementation depends on them being the default.

Reply:
I use the same comparer that the original Linq implementation used (unless I'm mistaken). Let me know what you want me to verify.

Reply:
That `left.Comparer` == `right.Comparer` == `EqualityComparer<object>.Default`.  If those differ, this routine may not do the correct thing.

Reply:
I only use `EqualityComparer<object>.Default` to compare the values, not for the dictionary. I did not change the comparer of the dictionary.

Reply:
I understand that.  I'm saying that you've added a utility method here that depends other code creating the dictionaries in a certain way in order for this code to work correctly.  As such this code should assert they were created correctly.